File: services/extraction-service/app/infrastructure/messaging/rabbitmq_publisher.py
import os
import json
import uuid
from datetime import datetime, timezone
import aio_pika
import structlog

logger = structlog.get_logger("extraction-service")

class RabbitMQPublisher:
    def __init__(self):
        self.rabbitmq_url = os.getenv("RABBITMQ_URL", "amqp://guest:guest@localhost:5672/")
        self.exchange_name = "extraction.exchange"
        self.connection = None
        self.channel = None
        self.exchange = None

    async def connect(self):
        logger.info("Connecting to RabbitMQ for publishing", url=self.rabbitmq_url)
        self.connection = await aio_pika.connect_robust(self.rabbitmq_url)
        self.channel = await self.connection.channel()
        self.exchange = await self.channel.declare_exchange(
            self.exchange_name, 
            aio_pika.ExchangeType.TOPIC, 
            durable=True
        )
        logger.info("Publisher connected to RabbitMQ")

    async def close(self):
        if self.connection:
            await self.connection.close()

    async def publish_extraction_completed(
        self, tenant_id: str, document_id: str, extraction_id: str, confidence_score: float, provider: str
    ):
        if not self.exchange:
            await self.connect()

        routing_key = "extraction.completed"
        
        event_payload = {
            "eventId": str(uuid.uuid4()),
            "eventType": "ExtractionCompletedEvent",
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "tenantId": tenant_id,
            "payload": {
                "documentId": document_id,
                "extractionId": extraction_id,
                "status": "SUCCESS",
                "confidenceScore": confidence_score,
                "providerUsed": provider
            }
        }

        message = aio_pika.Message(
            body=json.dumps(event_payload).encode(),
            delivery_mode=aio_pika.DeliveryMode.PERSISTENT,
            content_type="application/json",
            timestamp=datetime.now(timezone.utc)
        )

        logger.info("Publishing ExtractionCompletedEvent", document_id=document_id, extraction_id=extraction_id)
        await self.exchange.publish(message, routing_key=routing_key)

File: services/validation-service/src/infrastructure/messaging/RabbitMQPublisher.ts
import { EventPublisher } from '../../application/ports/EventPublisher.port';
import { DomainEvent } from '@distill/types';
import amqplib from 'amqplib';

export class RabbitMQPublisher implements EventPublisher {
  private connection: amqplib.Connection | null = null;
  private channel: amqplib.Channel | null = null;

  constructor(
    private readonly rabbitUrl: string,
    private readonly logger: any
  ) {}

  async connect(): Promise<void> {
    if (this.connection) return;

    try {
      this.connection = await amqplib.connect(this.rabbitUrl);
      this.channel = await this.connection.createConfirmChannel();

      // Assert validation exchange
      await this.channel.assertExchange('validation.exchange', 'topic', { durable: true });
      
      this.logger.info('RabbitMQ Publisher connected');
    } catch (error) {
      this.logger.error({ error }, 'Failed to connect to RabbitMQ');
      throw error;
    }
  }

  async publish(exchange: string, routingKey: string, event: DomainEvent): Promise<boolean> {
    if (!this.channel) {
      throw new Error('RabbitMQ channel not initialized. Call connect() first.');
    }

    const payload = Buffer.from(JSON.stringify(event));

    return new Promise((resolve, reject) => {
      this.channel!.publish(
        exchange,
        routingKey,
        payload,
        {
          persistent: true,
          contentType: 'application/json',
          messageId: event.eventId,
          timestamp: new Date(event.timestamp).getTime(),
          appId: 'validation-service',
          headers: {
            'x-tenant-id': event.tenantId
          }
        },
        (err) => {
          if (err) {
            this.logger.error({ error: err, eventId: event.eventId }, 'Failed to publish message');
            reject(err);
          } else {
            resolve(true);
          }
        }
      );
    });
  }

  async close(): Promise<void> {
    if (this.channel) await this.channel.close();
    if (this.connection) await this.connection.close();
    this.logger.info('RabbitMQ Publisher disconnected');
  }
}
